core/pkg/artifacts/saver.go

```diff
@@ -144,99 +153,165 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
+			manifest.Contents[fileInfo.Name] = entry
+			as.cacheEntry(entry)
+			if fileInfo.UploadUrl == nil {
+				// The server already has this file.
+				numActive--
+				as.numDone++
 				continue
 			}
-			numDone++
-			entry := manifest.Contents[result.Name]
-			if !entry.SkipCache {
-				digest := entry.Digest
-				go func() {
-					err := as.FileCache.AddFileAndCheckDigest(result.Task.Path, digest)
-					if err != nil {
-						slog.Error("error adding file to cache", "err", err)
-					}
-				}()
+			task := newUploadTask(fileInfo, *entry.LocalPath)
+			task.SetCompletionCallback(func(t *filetransfer.Task) { doneChan <- t })
+			as.FileTransferManager.AddTask(task)
+		// Listen for completed uploads, adding to the retry list if they failed.
+		case result := <-doneChan:
+			numActive--
+			if result.Err != nil {
+				mustRetry[result.Name] = namedFileSpecs[result.Name]
+			} else {
+				as.numDone++
 			}
+		// Check for errors.
+		case err := <-errorChan:
+			return nil, err
 		}
 	}
-	return nil
+	return mustRetry, nil
+}
+
+// batchFileDataRetriever takes a batch of file specs, requests upload URLs for each file,
+// assembles the info needed for the next step, and feeds them into an output channel.
+func (as *ArtifactSaver) batchFileDataRetriever(
+	batch []gql.CreateArtifactFileSpecInput,
+	resultChan chan<- serverFileResponse,
+	errorChan chan<- error,
+) {
+	response, err := gql.CreateArtifactFiles(
+		as.Ctx, as.GraphqlClient, batch, gql.ArtifactStorageLayoutV2,
+	)
+	if err != nil {
+		errorChan <- fmt.Errorf("requesting upload URLs failed: %v", err)
+		return
+	}
+	batchDetails := response.CreateArtifactFiles.Files.Edges
+	if len(batch) != len(batchDetails) {
+		errorChan <- fmt.Errorf("expected %v upload URLs, got %v", len(batch), len(batchDetails))
+		return
+	}
+	for i, edge := range batchDetails {
+		resultChan <- serverFileResponse{
+			Name:            batch[i].Name,
+			BirthArtifactID: edge.Node.Artifact.Id,
+			UploadUrl:       edge.Node.UploadUrl,
+			UploadHeaders:   edge.Node.UploadHeaders,
+		}
+	}
+}
+
+func (as *ArtifactSaver) nextBatch(
+	fileSpecs []gql.CreateArtifactFileSpecInput,
+) ([]gql.CreateArtifactFileSpecInput, []gql.CreateArtifactFileSpecInput) {
+	batchSize := min(as.batchSize(), len(fileSpecs))
+	return fileSpecs[:batchSize], fileSpecs[batchSize:]
+}
+
+func (as *ArtifactSaver) batchSize() int {
+	// We want to keep the number of pending uploads under the concurrency limit until
+	// we know how fast they upload.
+	minBatchSize := filetransfer.DefaultConcurrencyLimit / as.maxActiveBatches
+	maxBatchSize := 10000 / as.maxActiveBatches
+	if as.numDone < filetransfer.DefaultConcurrencyLimit {
+		return minBatchSize
+	}
+	// Given the average time per item, estimate a batch size that will take 1 minute.
+	sinceStart := time.Since(as.startTime)
+	filesPerMin := int(float64(as.numDone) / sinceStart.Minutes())
+	return max(min(maxBatchSize, filesPerMin), minBatchSize)
```

Comment:
I think this mostly makes sense, will we run into issues when `maxActivebatches` is low like 1 and we don't end up saturating `TransferManager` due to waiting for signed urls with a small batch size and this creates a feedback loop where we never up the batch size?

Reply:
Good question! I don't think it's possible given the current settings, since with the minimum batch size we still maintain a queue of 125 uploads ready to go. I certainly don't see this as in issue so far in my tests.